meta/autoid/autoid.go

```diff
@@ -348,3 +351,142 @@ var localSchemaID = int64(math.MaxInt64)
 func GenLocalSchemaID() int64 {
 	return atomic.AddInt64(&localSchemaID, -1)
 }
+
+// AllocN implements autoid.Allocator Alloc interface.
+func (alloc *allocator) AllocN(tableID int64, N uint64) ([]int64, error) {
+	if tableID == 0 {
+		return nil, errInvalidTableID.GenWithStackByArgs("Invalid tableID")
+	}
+	if N == 0 {
+		return []int64{}, nil
+	}
+	alloc.mu.Lock()
+	defer alloc.mu.Unlock()
+	if alloc.isUnsigned {
+		return alloc.allocN4Unsigned(tableID, N)
+	}
+	return alloc.allocN4Signed(tableID, N)
+}
+
+func (alloc *allocator) allocN4Signed(tableID int64, N uint64) ([]int64, error) {
+	N1 := int64(N)
+	// Condition alloc.base+N1 > alloc.end will overflow when alloc.base + N1 > MaxInt64. So need this.
+	if math.MaxInt64-alloc.base < N1 {
+		return nil, ErrAutoincReadFailed
+	}
+	// The local rest is not enough for allocN, skip it.
+	if alloc.base+N1 > alloc.end {
+		var newBase, newEnd int64
+		startTime := time.Now()
+		// Although it may skip a segment here, we still think it is consumed.
+		consumeDur := startTime.Sub(alloc.lastAllocTime)
+		nextStep := NextStep(alloc.step, consumeDur)
+		// Make sure nextStep is big enough.
+		if nextStep <= N1 {
+			alloc.step = mathutil.MinInt64(N1*2, maxStep)
+		} else {
+			alloc.step = nextStep
+		}
+		err := kv.RunInNewTxn(alloc.store, true, func(txn kv.Transaction) error {
+			m := meta.NewMeta(txn)
+			var err1 error
+			newBase, err1 = m.GetAutoTableID(alloc.dbID, tableID)
+			if err1 != nil {
+				return err1
+			}
+			tmpStep := mathutil.MinInt64(math.MaxInt64-newBase, alloc.step)
+			// The global rest is not enough for allocN.
+			if tmpStep < N1 {
+				return ErrAutoincReadFailed
+			}
+			newEnd, err1 = m.GenAutoTableID(alloc.dbID, tableID, tmpStep)
+			return err1
+		})
+		metrics.AutoIDHistogram.WithLabelValues(metrics.TableAutoIDAlloc, metrics.RetLabel(err)).Observe(time.Since(startTime).Seconds())
+		if err != nil {
+			return nil, err
+		}
+		alloc.lastAllocTime = time.Now()
+		if newBase == math.MaxInt64 {
+			return nil, ErrAutoincReadFailed
+		}
+		alloc.base, alloc.end = newBase, newEnd
+	}
+	logutil.BgLogger().Debug("alloc N signed ID",
```

Comment:
the new logger format is `logutil.Logger(ctx).Info` like~

Reply:
Nice!